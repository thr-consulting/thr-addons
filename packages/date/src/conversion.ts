import {LocalDate, LocalDateTime, nativeJs, ZonedDateTime, ZoneId, LocalTime} from '@js-joda/core';

/*
  Possible Date Types

  LocalDate - js-joda date with no time
  Object - Object with fields that look like LocalDate internals. See ILocalDateLike below.
  Epoch Days - js-joda number of days since epoch (Jan 1, 1970)
  LocalDateTime - js-joda date with time (no zone)
  ZonedDateTime - js-joda date with time and zone
  Date - javascript date with time (no zone)
  String - ISO8601 date string (without time, with time, with time and zone)
  LocalTime - js-joda time (no zone)
 */

export interface ILocalDateLike {
	year: number;
	month: number;
	day: number;
}

export function isLocalDateLike(date: any): date is ILocalDateLike {
	return (
		date !== null &&
		date !== undefined &&
		typeof date === 'object' &&
		typeof date.year !== 'undefined' &&
		typeof date.month !== 'undefined' &&
		typeof date.day !== 'undefined'
	);
}

export interface ILocalTimeLike {
	hour: number;
	minute: number;
	second: number;
	nano: number;
}

export function isLocalTimeLike(time: any): time is ILocalTimeLike {
	return (
		time !== null &&
		time !== undefined &&
		typeof time === 'object' &&
		typeof time.hour !== 'undefined' &&
		typeof time.minute !== 'undefined' &&
		typeof time.second !== 'undefined' &&
		typeof time.nano !== 'undefined'
	);
}

// const iso8601Regex = /^([+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24:?00)([.,]\d+(?!:))?)?(\17[0-5]\d([.,]\d+)?)?([zZ]|([+-])([01]\d|2[0-3]):?([0-5]\d)?)?)?)?$/;
const iso8601DateOnly = /^\d{4}-\d{2}-\d{2}$/;
const iso8601ContainsZone = /(Z|[+-]\d{2}:\d{2})$/;
const iso8601TimeOnly = /^\d{2}:\d{2}(:\d{2}(\.\d{1,9})?)?$/;

export function toLocalDate(date: any, zone: ZoneId = ZoneId.SYSTEM): LocalDate {
	if (date instanceof LocalDate) {
		return date;
	}
	if (date instanceof LocalDateTime) {
		// This chops the time off, because 'local' datetime === 'local date.
		return date.toLocalDate();
	}
	if (date instanceof ZonedDateTime) {
		// Unsure of this
		return date.toLocalDate();
	}
	if (typeof date === 'number') {
		return LocalDate.ofEpochDay(date);
	}
	if (date instanceof Date) {
		// Javascript dates are evil. They assume you are in SYSTEM zone when reading, but can set as any zone.
		return LocalDate.from(nativeJs(date, zone));
	}
	if (isLocalDateLike(date)) {
		// eslint-disable-next-line no-underscore-dangle
		return LocalDate.of(date.year, date.month, date.day);
	}
	if (typeof date === 'string') {
		if (iso8601DateOnly.test(date)) {
			return LocalDate.parse(date);
		}
		if (iso8601ContainsZone.test(date)) {
			return ZonedDateTime.parse(date).toLocalDate();
		}
		return LocalDateTime.parse(date).toLocalDate();
	}
	throw new Error('Cannot convert value to LocalDate');
}

export function toEpochDay(date: any, zone: ZoneId = ZoneId.SYSTEM): number {
	if (typeof date === 'number') {
		return date;
	}
	if (date instanceof LocalDate) {
		return date.toEpochDay();
	}
	if (date instanceof LocalDateTime || date instanceof ZonedDateTime || date instanceof Date || isLocalDateLike(date) || typeof date === 'string') {
		return toLocalDate(date, zone).toEpochDay();
	}

	throw new Error('Cannot convert value to epoch integer');
}

export function toLocalDateTime(date: any, zone: ZoneId = ZoneId.SYSTEM): LocalDateTime {
	if (date instanceof LocalDateTime) {
		return date;
	}
	if (date instanceof ZonedDateTime) {
		return date.toLocalDateTime();
	}
	if (date instanceof Date) {
		return LocalDateTime.from(nativeJs(date, zone));
	}
	if (date instanceof LocalDate || isLocalDateLike(date) || typeof date === 'number') {
		const localDate = toLocalDate(date);
		return LocalDateTime.of(localDate.year(), localDate.monthValue(), localDate.dayOfMonth(), 0, 0, 0, 0);
	}
	if (typeof date === 'string') {
		if (iso8601DateOnly.test(date)) {
			return toLocalDateTime(LocalDate.parse(date));
		}
		if (iso8601ContainsZone.test(date)) {
			return ZonedDateTime.parse(date).toLocalDateTime();
		}
		return LocalDateTime.parse(date);
	}
	throw new Error('Cannot convert value to LocalDateTime');
}

export function toDate(date: any): Date {
	if (date instanceof Date) {
		return date;
	}
	if (date instanceof LocalDate || date instanceof ZonedDateTime) {
		return new Date(date.year(), date.monthValue() - 1, date.dayOfMonth());
	}
	if (date instanceof LocalDateTime) {
		return new Date(date.year(), date.monthValue() - 1, date.dayOfMonth(), date.hour(), date.minute(), date.second(), date.nano() / 1000000);
	}
	if (date instanceof LocalTime) {
		return new Date(1970, 1, 1, date.hour(), date.minute(), date.second(), date.nano() / 1000000);
	}
	if (typeof date === 'number' || isLocalDateLike(date)) {
		return toDate(toLocalDate(date));
	}
	if (typeof date === 'string') {
		return new Date(date);
	}
	throw new Error('Cannot convert value to Date');
}

export function toLocalTime(time: any, zone: ZoneId = ZoneId.SYSTEM): LocalTime {
	if (time instanceof LocalTime) {
		return time;
	}
	if (time instanceof LocalDateTime) {
		return time.toLocalTime();
	}
	if (time instanceof ZonedDateTime) {
		return time.toLocalTime();
	}
	if (time instanceof Date) {
		return LocalTime.from(nativeJs(time, zone));
	}
	if (isLocalTimeLike(time)) {
		// eslint-disable-next-line no-underscore-dangle
		return LocalTime.of(time.hour, time.minute, time.second, time.nano);
	}
	if (typeof time === 'number') {
		return LocalTime.ofSecondOfDay(time);
	}
	if (typeof time === 'string') {
		if (iso8601TimeOnly.test(time)) {
			return LocalTime.parse(time);
		}
		return toLocalDateTime(time, zone).toLocalTime();
	}
	throw new Error('Cannot convert value to LocalTime');
}
