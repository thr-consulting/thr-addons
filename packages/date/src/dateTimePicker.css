@import "../node_modules/react-widgets/dist/css/react-widgets.css";

.rw-datetimepicker > input {
	border: none !important;
}

.rw-datetimepicker .rw-calendar-grid.rw-nav-view .rw-btn {
	padding: 0 !important;
}

.dateTimePickerAutoWidth {
	width: auto;
	display: inline-flex;
}

.dateTimePickerAutoWidth > input {
	width: auto;
}
