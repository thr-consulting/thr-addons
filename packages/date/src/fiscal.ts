import {LocalDate, Period} from '@js-joda/core';

export enum PeriodEnum {
	'Yearly' = 'yearly',
	'Quarterly' = 'quarterly',
}

export function PeriodEnumLookup(value: string) {
	switch (value) {
		case 'yearly':
			return PeriodEnum.Yearly;
		case 'quarterly':
			return PeriodEnum.Quarterly;
		default:
			throw new Error('Cannot find enum value: PeriodEnumLookup');
	}
}

export interface FiscalDateRange {
	start: LocalDate;
	end: LocalDate;
}

/**
 * Returns the @js-joda/Period from a PeriodEnum
 * @param period
 */
export function getFiscalPeriod(period: PeriodEnum): Period {
	switch (period) {
		case PeriodEnum.Yearly:
			return Period.of(1, 0, 0);
		case PeriodEnum.Quarterly:
			return Period.of(0, 3, 0);
		default:
			throw new Error(`Invalid period: ${period}`);
	}
}

/**
 * Utility to get the correct year-end date for a target year, handling Feb 29.
 */
function safeYearEndForYear(yearEnd: LocalDate, targetYear: number): LocalDate {
	const month = yearEnd.monthValue();
	const day = yearEnd.dayOfMonth();

	// If the year-end is Feb 29, return Feb 28 if target year is not a leap year.
	if (month === 2 && day === 29) {
		if (!LocalDate.of(targetYear, 1, 1).isLeapYear()) {
			return LocalDate.of(targetYear, 2, 28);
		}
	}
	return yearEnd.withYear(targetYear);
}

/**
 * Returns the fiscal year of a date.
 * The fiscal year is the year in which the fiscal period ends.
 * @param date
 * @param yearEnd
 */
export function getFiscalYear(date: LocalDate, yearEnd: LocalDate): number {
	const yearEndInDateYear = safeYearEndForYear(yearEnd, date.year());

	// If the date is strictly after the year end in its own calendar year, it belongs to the next fiscal year.
	if (date.isAfter(yearEndInDateYear)) {
		return date.year() + 1;
	}

	// Otherwise, the fiscal year ends in the current calendar year.
	return date.year();
}

/**
 * Returns the start and end dates for the date's year of fiscal year.
 * @param date
 * @param yearEnd
 */
export function getFiscalYearRange(date: LocalDate, yearEnd: LocalDate): FiscalDateRange {
	const endYear = getFiscalYear(date, yearEnd);
	const startYear = endYear - 1;
	const yearEndInStartYear = safeYearEndForYear(yearEnd, startYear);
	const startDate = yearEndInStartYear.plusDays(1);
	const endDate = safeYearEndForYear(yearEnd, endYear);
	return {
		start: startDate,
		end: endDate,
	};
}

/**
 * Returns the quarter of a fiscal year for the date.
 * @param date
 * @param yearEnd
 */
export function getFiscalQuarter(date: LocalDate, yearEnd: LocalDate): 1 | 2 | 3 | 4 {
	const {start} = getFiscalYearRange(date, yearEnd);

	// 3-month quarter checks:
	// Q1: date < start + 3 months
	if (date.isBefore(start.plusMonths(3))) return 1;
	// Q2: date < start + 6 months
	if (date.isBefore(start.plusMonths(6))) return 2;
	// Q3: date < start + 9 months
	if (date.isBefore(start.plusMonths(9))) return 3;
	// Q4: otherwise
	return 4;
}

/**
 * Returns the start and end dates of the date's fiscal quarter.
 * @param date
 * @param yearEnd
 */
export function getFiscalQuarterRange(date: LocalDate, yearEnd: LocalDate): FiscalDateRange {
	const {start} = getFiscalYearRange(date, yearEnd);
	const q = getFiscalQuarter(date, yearEnd);
	const quarterStart = start.plusMonths(3 * (q - 1));
	const quarterEnd = quarterStart.plusMonths(3).minusDays(1);

	return {
		start: quarterStart,
		end: quarterEnd,
	};
}

/**
 * Returns the start and end dates of the date's fiscal range depending on period.
 * @param period
 * @param date
 * @param yearEnd
 */
export function getFiscalRange(period: PeriodEnum, date: LocalDate, yearEnd: LocalDate): FiscalDateRange {
	if (period === PeriodEnum.Yearly) {
		return getFiscalYearRange(date, yearEnd);
	}

	return getFiscalQuarterRange(date, yearEnd);
}
