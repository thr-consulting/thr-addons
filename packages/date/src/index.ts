export {formatDate, FormatDateType} from './formatDate';
export {getMonthNames} from './getMonthNames';
export {getNumberOfDaysInMonth} from './getNumberOfDaysInMonth';
export {toLocalDate, toLocalDateTime, toDate, toEpochDay, toLocalTime, toSecondOfDay} from './conversion';
export {getTimeOfDay} from './specialFormats';
export {
	getFiscalYear,
	getFiscalPeriod,
	getFiscalYearRange,
	getFiscalQuarterRange,
	getFiscalQuarter,
	getFiscalRange,
	shiftFiscalRangeToYear,
	safeYearEndForYear,
	PeriodEnumLookup,
	PeriodEnum,
} from './fiscal';
export type {FiscalDateRange} from './fiscal';
export type {FormatDateParams} from './formatDate';
