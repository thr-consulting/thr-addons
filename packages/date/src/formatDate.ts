import {DateTimeFormatter, LocalDate, LocalDateTime, LocalTime, ZonedDateTime, ZoneId} from '@js-joda/core';
// @ts-ignore
import {Locale} from '@js-joda/locale_en-us';
import {toLocalDate, toLocalDateTime, ILocalDateLike, isLocalDateLike} from './conversion';

export enum FormatDateType {
	short = 'short',
	medium = 'medium',
	long = 'long',
}

export interface FormatDateParams {
	type?: FormatDateType | 'short' | 'medium' | 'long';
	time?: boolean;
	date?: boolean;
	format?: string;
}

/**
 * Formats a date to a string
 * @param obj
 * @param params
 * @param zone
 */
export function formatDate(
	obj:
		| ILocalDateLike
		| Date
		| LocalDate
		| number
		| string
		| LocalDateTime
		| LocalTime
		| ZonedDateTime
		| null
		| undefined,
	params: FormatDateParams = {type: FormatDateType.short, time: false, date: true},
	zone: ZoneId = ZoneId.SYSTEM,
) {
	let l: LocalDate | LocalDateTime | LocalTime | ZonedDateTime | undefined;
	if (!obj) return '';
	if (typeof obj === 'number') l = toLocalDate(obj);
	if (isLocalDateLike(obj)) l = toLocalDate(obj);
	if (obj instanceof Date) l = toLocalDateTime(obj, zone);
	if (obj instanceof LocalDate) l = obj;
	if (obj instanceof LocalDateTime) l = obj;
	if (obj instanceof LocalTime) l = obj;
	if (obj instanceof ZonedDateTime) l = obj;
	if (!l) throw new Error('Date could not be converted to a js-joda date');

	let dateFormat: string;
	let timeFormat: string;
	switch (params.type) {
		case FormatDateType.medium:
			dateFormat = 'MMM d, yyyy';
			timeFormat = 'h:mm a';
			break;
		case FormatDateType.long:
			dateFormat = 'MMMM d, yyyy';
			timeFormat = 'h:mm a';
			break;
		case FormatDateType.short:
		default:
			dateFormat = 'M/d/yyyy';
			timeFormat = 'h:mm a';
			break;
	}

	if (l instanceof LocalDate) timeFormat = '';
	if (l instanceof LocalTime) dateFormat = '';

	let formatString = !params.date ? '' : dateFormat;
	if (params.time) {
		formatString = `${formatString} ${timeFormat}`;
	}

	const formatter = DateTimeFormatter.ofPattern(params.format || formatString).withLocale(Locale.ENGLISH);

	return l.format(formatter).replace(/AM|PM/, x => x.toLowerCase());
}
