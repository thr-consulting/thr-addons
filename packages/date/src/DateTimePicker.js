// @flow

import React from 'react';
import DTPicker from 'react-widgets/lib/DateTimePicker';
import './dateTimePicker.css';

type Props = {
	width?: string,
};

/**
 * DateTimePicker from {@link https://jquense.github.io/react-widgets/docs/#/datetime-picker|React Widgets}.
 * @class
 * @property {string} [width=null] - If set to 'auto' the input fills the width.
 * @property {Date} value - The JS Date object
 * @property {onChange} onChange - Called when the value changes.
 */
export default function DateTimePicker(props: Props) {
	let extraClasses;
	if (props.width === 'auto') {
		extraClasses = 'dateTimePickerAutoWidth';
	}
	const aprops = {...props};
	delete aprops.meta;
	return <DTPicker {...aprops} className={extraClasses}/>;
}
