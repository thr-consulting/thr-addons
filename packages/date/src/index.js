// @flow

import DateTimePicker from './DateTimePicker';
import LocalDatePicker from './LocalDatePicker';
import MonthSelect from './MonthSelect';
import LocalMonthSelect from './LocalMonthSelect';
import {
	formatDate,
	transformDatesToMoment,
	transformMomentsToDate,
	transformDateToLocalDate,
	transformLocalDateToDate,
	transformLocalDateToMoment,
	transformMomentToLocalDate,
	transformLocalDatesToEpochInteger,
	transformEpochIntegerToLocalDate,
	mapEpochIntegerToLocalDates,
	transformObjectsToLocalDates,
	dateInit,
} from './util';

export {
	DateTimePicker,
	LocalDatePicker,
	MonthSelect,
	LocalMonthSelect,
	dateInit,
	formatDate,
	transformMomentsToDate,
	transformDatesToMoment,
	transformDateToLocalDate,
	transformLocalDateToDate,
	transformLocalDateToMoment,
	transformMomentToLocalDate,
	transformLocalDatesToEpochInteger,
	transformEpochIntegerToLocalDate,
	mapEpochIntegerToLocalDates,
	transformObjectsToLocalDates,
};
