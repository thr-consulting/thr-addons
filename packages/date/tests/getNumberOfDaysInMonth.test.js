import {advanceTo, clear} from 'jest-date-mock';
import {getNumberOfDaysInMonth} from '../src';

describe('getNumberOfDaysInMonth', () => {
	it('should return the number 31 for January', () => {
		expect(getNumberOfDaysInMonth({month: 1})).toMatchSnapshot();
	});

	it('should return the number 28 for February', () => {
		expect(getNumberOfDaysInMonth({month: 2})).toMatchSnapshot();
	});

	it('should return the number 30 for April', () => {
		expect(getNumberOfDaysInMonth({month: 4})).toMatchSnapshot();
	});

	it('should return the number 13 for this month', () => {
		advanceTo(new Date(2019, 5, 13, 0, 0, 0));
		expect(getNumberOfDaysInMonth({month: 6, limitToCurrentDay: true})).toMatchSnapshot();
		clear();
	});

	it('should return an array of 30 for this month', () => {
		advanceTo(new Date(2019, 5, 13, 0, 0, 0));
		expect(getNumberOfDaysInMonth({month: 6, asArray: true})).toMatchSnapshot();
	});

	it('should return an array of 5 for this month', () => {
		advanceTo(new Date(2019, 5, 5, 0, 0, 0));
		expect(getNumberOfDaysInMonth({month: 6, limitToCurrentDay: true, asArray: true})).toMatchSnapshot();
	});
});
