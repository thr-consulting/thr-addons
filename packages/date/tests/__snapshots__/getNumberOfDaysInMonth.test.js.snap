// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNumberOfDaysInMonth should return an array of 5 for this month 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`getNumberOfDaysInMonth should return an array of 30 for this month 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
  25,
  26,
  27,
  28,
  29,
  30,
]
`;

exports[`getNumberOfDaysInMonth should return the number 13 for this month 1`] = `13`;

exports[`getNumberOfDaysInMonth should return the number 28 for February 1`] = `28`;

exports[`getNumberOfDaysInMonth should return the number 30 for April 1`] = `30`;

exports[`getNumberOfDaysInMonth should return the number 31 for January 1`] = `31`;
